lib/lib-storage/src/s3-transfer-manager/transfer-manager-types.ts

```diff
@@ -0,0 +1,331 @@
+   *
+   * @param request - All properties of a single or multipart upload request.
+   * @param options - Allows users to specify cancel functions for the request.
+   * @param eventListeners - Collection of callbacks for monitoring transfer lifecycle events. Allows tracking statuses per request.
+   *
+   * @returns The response from the S3 API for the upload request.
+   */
+  upload(
+    request: UploadRequest,
+    options?: HttpHandlerOptions,
+    eventListeners?: TransferEventListeners
+  ): Promise<UploadResponse>;
+
+  /**
+   * Lets users download single objects from a given bucket to a given directory.
+   * Supports multipart download, single object download, and transfer progress listeners.
+   *
+   * @param request - All properties of a single or multipart upload request.
+   * @param options - Allows users to specify cancel functions for the request.
+   * @param eventListeners - Collection of callbacks for monitoring transfer lifecycle events. Allows tracking statuses per request.
+   *
+   * @returns The response from the S3 API for the download request.
+   */
+  download(
+    request: DownloadRequest,
+    options?: HttpHandlerOptions,
+    eventListeners?: TransferEventListeners
+  ): Promise<DownloadResponse>;
+
+  /**
+   * Represents an API to upload all files under the given directory to the provided S3 bucket.
+   *
+   * @param options.bucket - The name of the bucket to upload objects to.
+   * @param options.source - The source directory to upload.
+   * @param options.followSymbolicLinks - Whether to follow symbolic links when traversing the file tree.
+   * @param options.recursive - Whether to upload directories recursively.
+   * @param options.s3Prefix - The S3 key prefix to use for each object. If not provided, files will be uploaded to the root of the bucket todo()
+   * @param options.s3Delimiter - Default "/". The S3 delimiter. A delimiter causes a list operation to roll up all the keys that share a common prefix into a single summary list result.
+   * @param options.putObjectRequestCallback - A callback mechanism to allow customers to update individual putObjectRequest that the S3 Transfer Manager generates.
+   * @param options.failurePolicy - The failure policy to handle failed requests
+   * @param eventListeners - Collection of callbacks for monitoring transfer lifecycle events. Allows tracking statuses of directory uploads.
+   *
+   * @returns The number of objects that have been uploaded and the number of objects that have failed
+   */
+  uploadAll(options: {
+    bucket: string;
+    source: string;
+    followSymbolicLinks?: boolean;
+    recursive?: boolean;
+    s3Prefix?: string;
+    filter?: (filepath: string) => boolean;
+    s3Delimiter?: string;
+    putObjectRequestCallback?: (putObjectRequest: PutObjectCommandInput) => Promise<void>;
+    failurePolicy?: (error?: unknown) => Promise<void>;
+    eventListeners?: TransferEventListeners;
+  }): Promise<{
```

Comment:
the recursive methods need to accept AbortSignals, probably via HttpHandlerOptions as well.